borsh-derive-internal/src/attribute_helpers.rs

```diff
@@ -18,3 +23,120 @@ pub fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
 
     None
 }
+pub fn check_use_discriminant(derive_input: DeriveInput) -> Result<Option<bool>, TokenStream> {
```

Comment:
this method tried to do too many different things at once. (imho)
let's split it in two. 
First: 
```rust
pub fn check_item_attributes(derive_input: &DeriveInput) -> Result<(), TokenStream> {
```
It should be at the same spot, as `check_use_discriminant` is now in `borsh-derive/src/lib.rs` similarly to the following:
```rust
check_item_attributes(&derive_input)?;
```
It should check that `borsh_skip` is not used at item level,  `borsh(attr)` only uses the `USE_DISCRIMINANT` allowed atm in place of `attr`, and that `USE_DISCRIMINANT` is only used for enum, not struct.

Second: 
```rust
contains_use_discriminant(attrs: &[Attribute]) -> Result<Option<bool>, TokenStream>
```
should be used from within `enum_ser` and `enum_de` methods similarly to 
```rust
let init_method = contains_initialize_with(&input.attrs);
```
 with these methods' (`enum_ser` and `enum_de`) signatures unchanged:
```rust
let use_discriminant: Option<bool> = contains_use_discriminant(&input.attrs)?;
```
One of reasons for this change is that it'll also simplify rebase and conflicts fixing.